extension/storage/storage.go

```diff
@@ -0,0 +1,102 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package storage
+
+import (
+	"context"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/config"
+)
+
+// Extension is the interface that storage extensions must implement
+type Extension interface {
+	component.Extension
+
+	// GetClient will create a client for use by the specified component.
+	// The component can use the client to manage state
+	GetClient(context.Context, component.Kind, config.ComponentID, string) (Client, error)
```

Comment:
Can we add parameter names to make clear what they are? The last `string` is particularly unclear.

Reply:
I described the last param as "signalName", since that was the initial intention, though it can be more generic and describe any unique client for a given component. I didn't want to make it too complex and left it as "signalName" but perhaps we should be more general about it?

Reply:
Yes, I think something more generic would be more suitable. Perhaps `storageName` or `substorageName` and then explain in the comments that each component can have multiple storages and that's why we have storage names?

Reply:
Yeah, that's a good idea. Updated it with `storageName`